File: essencium-backend-development/src/main/java/de/frachtwerk/essencium/backend/service/UserService.java
/*
 * Copyright (C) 2025 Frachtwerk GmbH, Leopoldstraße 7C, 76133 Karlsruhe.
 *
 * This file is part of essencium-backend.
 *
 * essencium-backend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * essencium-backend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with essencium-backend. If not, see <http://www.gnu.org/licenses/>.
 */

package de.frachtwerk.essencium.backend.service;

import de.frachtwerk.essencium.backend.model.Role;
import de.frachtwerk.essencium.backend.model.User;
import de.frachtwerk.essencium.backend.model.dto.AppUserDto;
import de.frachtwerk.essencium.backend.model.dto.EssenciumUserDetails;
import de.frachtwerk.essencium.backend.repository.UserRepository;
import jakarta.validation.constraints.NotNull;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class UserService
    extends AbstractUserService<User, EssenciumUserDetails<Long>, Long, AppUserDto> {

  @Autowired
  protected UserService(
      @NotNull UserRepository userRepository,
      @NotNull PasswordEncoder passwordEncoder,
      @NotNull UserMailService userMailService,
      @NotNull RoleService roleService,
      @NotNull AdminRightRoleCache adminRightRoleCache,
      @NotNull JwtTokenService jwtTokenService) {
    super(
        userRepository,
        passwordEncoder,
        userMailService,
        roleService,
        adminRightRoleCache,
        jwtTokenService);
  }

  @Override
  protected @NotNull <E extends AppUserDto> User convertDtoToEntity(
      @NotNull E entity, Optional<User> currentEntityOpt) {
    Set<Role> roles =
        entity.getRoles().stream().map(roleService::getByName).collect(Collectors.toSet());

    return User.builder()
        .email(entity.getEmail())
        .enabled(entity.isEnabled())
        .roles(roles)
        .firstName(entity.getFirstName())
        .lastName(entity.getLastName())
        .locale(entity.getLocale())
        .mobile(entity.getMobile())
        .phone(entity.getPhone())
        .source(entity.getSource())
        .id(entity.getId())
        .loginDisabled(entity.isLoginDisabled())
        .build();
  }

  @NotNull
  @Override
  protected Pageable getAllPreProcessing(@NotNull final Pageable pageable) {
    Sort.Order nameSortOrder = pageable.getSort().getOrderFor("name");

    if (nameSortOrder == null) {
      return pageable;
    }

    List<Sort.Order> orders = pageable.getSort().stream().collect(Collectors.toList());

    int nameSortIndex = orders.indexOf(nameSortOrder);

    Sort.Order firstNameSortOrder = nameSortOrder.withProperty("firstName");
    Sort.Order lastNameSortOrder = nameSortOrder.withProperty("lastName");

    orders.remove(nameSortIndex);

    orders.add(nameSortIndex, firstNameSortOrder);
    orders.add(nameSortIndex + 1, lastNameSortOrder);

    return PageRequest.of(pageable.getPageNumber(), pageable.getPageSize(), Sort.by(orders));
  }

  @Override
  public AppUserDto getNewUser() {
    return new AppUserDto();
  }
}

File: essencium-backend-development/src/main/kotlin/de/frachtwerk/essencium/backend/Service/UserService.kt
package de.frachtwerk.essencium.backend.service

import de.frachtwerk.essencium.backend.model.User
import de.frachtwerk.essencium.backend.model.dto.EssenciumUserDetails
import de.frachtwerk.essencium.backend.model.dto.UserDto
import de.frachtwerk.essencium.backend.repository.UserRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class UserService(
    userRepository: UserRepository,
    passwordEncoder: PasswordEncoder,
    userMailService: UserMailService,
    private val roleService: RoleService,
    adminRightRoleCache: AdminRightRoleCache,
    jwtTokenService: JwtTokenService
) : AbstractUserService<User, EssenciumUserDetails<Long>, Long, UserDto>(
    userRepository,
    passwordEncoder,
    userMailService,
    roleService,
    adminRightRoleCache,
    jwtTokenService
) {

    override fun <E : UserDto> convertDtoToEntity(
        dto: E,
        currentEntityOpt: Optional<User>
    ): User {
        val roles = dto.roles.map { roleName -> roleService.getByName(roleName) }.toSet()

        val user = currentEntityOpt.orElseGet { User() }

        if (dto.id != null) user.id = dto.id


        user.email = dto.email
        user.isEnabled = dto.isEnabled
        user.firstName = dto.firstName
        user.lastName = dto.lastName
        user.locale = dto.locale
        user.setMobile(dto.mobile)
        user.setPhone(dto.phone)
        user.source = dto.source
        user.isLoginDisabled = dto.isLoginDisabled
        user.roles = roles

        return user
    }


    override fun selfUpdate(user: User, updateInformation: UserDto): User {
        user.setPhone(updateInformation.phone)
        user.setMobile(updateInformation.mobile)
        return super.selfUpdate(user, updateInformation)
    }

    override fun selfUpdatePermittedFields(): Set<String> =
        super.selfUpdatePermittedFields().toMutableSet().apply {
            add("phone")
            add("mobile")
        }

    override fun getAllPreProcessing(pageable: Pageable): Pageable {
        val nameSortOrder = pageable.sort.getOrderFor("name") ?: return pageable
        val orders = pageable.sort.toList().toMutableList()
        val idx = orders.indexOf(nameSortOrder)

        val firstNameSortOrder = nameSortOrder.withProperty("firstName")
        val lastNameSortOrder = nameSortOrder.withProperty("lastName")

        orders.removeAt(idx)
        orders.add(idx, firstNameSortOrder)
        orders.add(idx + 1, lastNameSortOrder)

        return PageRequest.of(pageable.pageNumber, pageable.pageSize, Sort.by(orders))
    }

    override fun getNewUser(): UserDto = UserDto()
}
